### ISSUE TITLE
assigning_clones: try to use dropped value

### ISSUE THREAD

**Issue by yhx-12243** [rust-lang/rust-clippy#13087]
### Summary

Consider we has a `mut x` and want to do `x = y.to_owned()`/`x = y.clone()`, but `x` is temporarily dropped, then we can't use `clone_into`/`clone_from`.

### Lint Name

assigning_clones

### Reproducer

I tried this code:

```rust
#![allow(dead_code)]
#![warn(clippy::assigning_clones)]

fn f(x: String) -> usize { x.len() }

fn g(mut x: String, y: &str) -> String {
    f(x);
    x = y.to_owned();
    x
}

fn main() {}
```

I saw this happen:

```
warning: assigning the result of `ToOwned::to_owned()` may be inefficient
 --> 1.rs:8:2
  |
8 |     x = y.to_owned();
  |     ^^^^^^^^^^^^^^^^ help: use `clone_into()`: `y.clone_into(&mut x)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assigning_clones
note: the lint level is defined here
 --> 1.rs:2:9
  |
2 | #![warn(clippy::assigning_clones)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^

warning: 1 warning emitted
```

I expected to see this happen:

&lt;nothing&gt;

### Version

```text
rustc 1.81.0-nightly (0c81f94b9 2024-07-10)
binary: rustc
commit-hash: 0c81f94b9a6207fb1fc080caa83584dea2d71fc6
commit-date: 2024-07-10
host: aarch64-apple-darwin
release: 1.81.0-nightly
LLVM version: 18.1.7
```

**Comment by lukaslueg** [comment]
Same problem with partial moves:

```rust
#![allow(dead_code)]
#![warn(clippy::assigning_clones)]

fn main() {
    let mut foobar: (String, String) = Default::default();

    drop(foobar.0);
    foobar.0 = foobar.1.to_owned(); // suggests: foobar.1.clone_into(&mut foobar.0)
}
```

**Comment by WieslerAA** [comment]
Same problem using references only.

```
#![allow(dead_code)]
#![warn(clippy::assigning_clones)]

fn main() {
    let mut test = String::new();

    let r: &mut String = &mut test;
    *r = r.trim().to_owned(); // suggests: r.trim().clone_into(r)
}
```

**Comment by y21** [comment]
@WieslerAA that case was fixed by #12756 and clippy shouldn't emit a warning for that anymore (on nightly for now, at least). Unfortunately it's not on stable yet, but will be part of the next release.

The reported issue in the OP is different and still happens on nightly

**Comment by samueltardieu** [comment]
It looks fixed since 1.79.0.

**Comment by yhx-12243** [comment]
> It looks fixed since 1.79.0.

No until now (1.83.0 nightly).
